flake8_trio/visitors/visitors.py

```diff
@@ -0,0 +1,415 @@
+            )
+        ) and not any(
+            node.name == opt
+            for opt in (*self.options.startable_in_context_manager, *STARTABLE_CALLS)
+        ):
+            self.error(node, node.name)
+
+
+# Suggests replacing all `trio.sleep(0)` with the more suggestive
+# `trio.lowlevel.checkpoint()`
+@error_class
+class Visitor115(Flake8TrioVisitor):
+    error_codes = {
+        "TRIO115": "Use `trio.lowlevel.checkpoint()` instead of `trio.sleep(0)`.",
+    }
+
+    def visit_Call(self, node: ast.Call):
+        if (
+            get_matching_call(node, "sleep")
+            and len(node.args) == 1
+            and isinstance(node.args[0], ast.Constant)
+            and node.args[0].value == 0
+        ):
+            self.error(node)
+
+
+@error_class
+class Visitor116(Flake8TrioVisitor):
+    error_codes = {
+        "TRIO116": (
+            "trio.sleep() with >24 hour interval should usually be "
+            "`trio.sleep_forever()`"
+        ),
+    }
+
+    def visit_Call(self, node: ast.Call):
+        if get_matching_call(node, "sleep") and len(node.args) == 1:
+            arg = node.args[0]
+            if (
+                # `trio.sleep(math.inf)`
+                (
+                    isinstance(arg, ast.Attribute)
+                    and isinstance(arg.value, ast.Name)
+                    and arg.attr == "inf"
+                    and arg.value.id == "math"
+                )
+                # `trio.sleep(inf)`
+                or (isinstance(arg, ast.Name) and arg.id == "inf")
+                # `trio.sleep(float("inf"))`
+                or (
+                    isinstance(arg, ast.Call)
+                    and isinstance(arg.func, ast.Name)
+                    and arg.func.id == "float"
+                    and len(arg.args)
+                    and isinstance(arg.args[0], ast.Constant)
+                    and arg.args[0].value == "inf"
+                )
+                # `trio.sleep(1e999)` (constant value inf)
+                # `trio.sleep(86401)`
+                # `trio.sleep(86400.1)`
+                or (
+                    isinstance(arg, ast.Constant)
+                    and isinstance(arg.value, (int, float))
+                    and arg.value > 86400
+                )
+            ):
+                self.error(node)
+
+
+@error_class
+@disabled_by_default
+class Visitor900(Flake8TrioVisitor):
+    error_codes = {
+        "TRIO900": ("Async generator without known-safe decorator not allowed.")
+    }
+
+    def __init__(self, *args: Any, **kwargs: Any):
+        super().__init__(*args, **kwargs)
+        self.unsafe_function: ast.AsyncFunctionDef | None = None
+
+    def visit_AsyncFunctionDef(
+        self, node: ast.AsyncFunctionDef | ast.FunctionDef | ast.Lambda
+    ):
+        self.save_state(node, "unsafe_function")
+        if isinstance(node, ast.AsyncFunctionDef) and not fnmatch_qualified_name(
+            node.decorator_list, *self.options.no_checkpoint_warning_decorators
```

Comment:
I think here that we should *only* exempt `asynccontextmanager`; as an optional check it's appropriate to be stricter and I really do want to get rid of all the existing non-context-manager cases I know about - they generally make it impossible to handle exceptions correctly, which includes cancellation and therefore timeouts.

Reply:
Addressing this in a separate PR.